src/db/tables/subscriber_coupons.ts

```diff
@@ -0,0 +1,54 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import createDbConnection from "../connect.js";
+import { logger } from "../../app/functions/server/logging";
+import { SubscriberCouponRow } from "knex/types/tables";
+
+const knex = createDbConnection();
+
+async function checkCouponForSubscriber(
+  subscriberId: number,
+  couponCode: string,
+) {
+  logger.info("checkCouponForSubscriber", subscriberId);
+  return (
+    (
+      await knex("subscriber_coupons")
+        .orderBy("created_at")
+        .where("subscriber_id", subscriberId)
+        .andWhere("coupon_code", couponCode)
+        .returning("*")
+    ).length > 0
+  );
+}
+
+async function addCouponForSubscriber(
+  subscriberId: number,
+  couponCode: string,
+) {
+  logger.info("addCouponForSubscriber", { subscriberId, couponCode });
+
+  let res;
+  try {
+    res = await knex("subscribers_coupon")
+      .insert({
+        subscriber_id: subscriberId,
+        coupon_code: couponCode,
+      })
+      .returning("*");
+  } catch (e) {
+    if ((e as Error).message.includes("violates unique constraint")) {
+      logger.info("addCouponForSubscriber", {
+        subscriberId,
+        error: (e as Error).message,
+      });
+    } else {
+      logger.error(e);
```

Comment:
This prints a pretty confusing message sometimes, I think it's better if we do something like:
```
logger.error("could_not_add_coupon", { error: JSON.stringify(e) });
```
Because when you see the message in the logs you'll just see whatever `e.toString()` returns, so it might not be clear exactly where it was thrown.

Do you think it should re-throw the error here? Otherwise the calling code won't have a way to know it didn't work, if that's important.

Reply:
Good idea to rethrow. I'll handle that on the caller side